packages/server/modules/shared/helpers/mocks.ts

```diff
@@ -0,0 +1,206 @@
+    return ref
+  }
+
+  const getMockRef = <T = any>(
+    type: string,
+    options?: {
+      /**
+       * The id of the object to get. If object w/ this ID already exists in the mock store,
+       * it will be retrieved from there. Otherwise, a new object will be created.
+       */
+      id?: string
+      /**
+       * Additional field values that should be set on the object
+       */
+      values?: Record<string, any>
+    }
+  ) => {
+    const { id, values } = options || {}
+    const ret = values
+      ? store.get(type, {
+          ...values,
+          ...(id ? { id } : {})
+        })
+      : store.get(type, id)
+    return ret as T
+  }
+
+  const getFieldValue = <T = any>(
+    refOrObj: Record<string, unknown> | Ref | SimpleRef,
+    field: string
+  ) => {
+    if (isRef(refOrObj)) return store.get(refOrObj, field) as T
+    if (isSimpleRef(refOrObj)) return store.get(refOrObj.type, refOrObj.id, field) as T
+    return refOrObj[field] as T
+  }
+
+  type AnyResolverFn = ResolverFn<any, any, any, any>
+
+  const resolveFromMockParent = (
+    options?: Partial<{
+      /**
+       * Allows you to map any refs found (whether they're in arrays or not) to something else,
+       * e.g. the same mock, but with different arg values
+       */
+      mapRefs: (
+        mockRef: Ref,
+        resolverArgs: { parent: any; args: any; ctx: any; info: GraphQLResolveInfo }
+      ) => any
+    }>
+  ) => {
+    const { mapRefs } = options || {}
+
+    const resolver: AnyResolverFn = (parent, args, ctx, info) => {
+      const resolverArgs = { parent, args, ctx, info }
+      const val = getFieldValue(parent, info.fieldName)
+      if (!mapRefs) return val
+
+      if (isArray(val)) {
+        return val.map((v) => (isRef(v) ? mapRefs(v, resolverArgs) : v))
+      } else {
+        return isRef(val) ? mapRefs(val, resolverArgs) : val
+      }
+    }
+
+    return resolver
+  }
+
+  const resolveAndCache = (resolver: AnyResolverFn) => {
+    const wrapperResolver: AnyResolverFn = (parent, args, ctx, info) => {
+      let cached: any
+
+      if (!isRef(parent) && !has(parent, 'id')) {
+        throw new Error(
+          'resolveAndCache depends on resolver parent being a mock ref or an object with an ID field'
+        )
+      }
+
+      if (isRef(parent)) {
+        if (hasField(parent.$ref.typeName, parent.$ref.key, info.fieldName)) {
+          cached = store.get(parent, info.fieldName)
+        }
+      } else {
+        if (hasField(info.parentType.name, parent.id, info.fieldName)) {
+          cached = store.get(info.parentType.name, parent.id, info.fieldName)
+        }
+      }
+
+      if (cached) return cached
+
+      const val = resolver(parent, args, ctx, info)
+      if (isRef(parent)) {
+        store.set(parent, info.fieldName, val)
+      } else {
+        store.set(info.parentType.name, parent.id, info.fieldName, val)
+      }
+
+      return val
+    }
+
+    return wrapperResolver
+  }
+
+  return {
```

Comment:
Pretty proud of these haha, these greatly simplify the entire mock definition experience and improve DX